README.md
![](example/screenshots/logo.gif)

## Features

- ✨ Simple Use
- ⚙️ Great Customization
- 🌍 Expand the functionality of your cards
- 🛑 Not use for mobile devices, will soon be available for use on those platforms.

## Getting Started

To use this card actions package, you need to add the dependencies in `pubspec.yaml` file.

```yaml
dependencies:
    card_actions: ^0.0.1
```

## Use

To use card actions is very simple, just wrap your **card** with our `CardActions` widget,

```dart
CardActions(
    ...
    child: YourCard() // here would be your card
)
```

### Basic Setup

**CardActions** needs the parameters `width`, `height` and `actions`

- `width` is the width of the widget
- `height` is the height of the widget
- `actions` is a list of `CardActionButton`.

```dart
CardActions(
    width: //your card width
    height: // your card height
    actions: <CardActionButton>[
        CardActionButton(
            icon: Icon(
                Icons.edit,
                color: Colors.white,
            ), // Icon
            label: 'edit',
            onPress: () {},
        ), // CardActionButton
    ]
    child: yourCard(), // here would be your card 
)
```

![](example/screenshots/first.gif)

You can add more buttons if you prefer

```dart
...
actions: [
    CardActionButton(
        icon: Icon(
            Icons.edit,
            color: Colors.white,
        ),
        label: 'edit',
        onPress: () {},
    ),
    CardActionButton(
        icon: Icon(
            Icons.payment_rounded,
            color: Colors.white,
        ),
        label: 'Pay',
        onPress: () {},
    ),
    CardActionButton(
        icon: Icon(
            Icons.remove_red_eye_sharp,
            color: Colors.white,
        ),
        label: 'Preview',
        onPress: () {},
    ),
]
```
![](example/screenshots/second.gif)

### Axis Direction

If you want to change the direction of the hover, you can use the `axisDirection` property of the `CardAction` to change it.

the `axisDirection` property receives a **enum** of type `CardActionAxis` at the moment we only have **bottom** and **right**.

| CardActionAxis.bottom | CardActionAxis.right     |
|-----------------------|--------------------------|
| ![](example/screenshots/second.gif) | ![](example/screenshots/horizontal.gif) |


## Example

![](example/screenshots/counter.gif)

To make this example, we will need to configure more things, but it will be very simple, the first thing is that we change the background color for it we have to use the property `backgroundColor` then an important part is that we change the design of the cursor when it is on top of the buttons and this we can do it in two ways:

- In the `CardAction` we have the `buttonsCursor` property that serves to change the cursor to **ALL** the buttons.

- In the `CardActionButton` we have the `cursor` property to change the cursor to that specific button.

> **Important**.
> If you use both options at the same time the button will prioritize its own `cursor` setting over the global setting with `buttonsCursor`.

and finally we add the necessary functions to the buttons by means of its `onPress` method.

```dart
...
actions: [
    CardActionButton(
        icon: Icon(
            Icons.add,
            color: Colors.white,
            size: 30,
        ),
        label: 'Increment',
        onPress: increment // method to increase the counter,
    ),
    CardActionButton(
        icon: Icon(
            Icons.restore_sharp,
            color: Colors.white,
            size: 30,
        ),
        label: 'Reset',
        onPress: () {
            // reset the counter
            setState(() {
            counter = 0;
            });
        },
    ),
    CardActionButton(
        icon: Icon(
            Icons.remove,
            color: Colors.white,
            size: 30,
        ),
        label: 'Decrement',
        onPress: decrement // method for decrementing the counter
    ),
]
```

## RoadMap
- Add options to handle card animation.
- Make it usable on mobile devices, but instead of hover, click on the card, or hold down.
- Add more axes and even be able to use several at the same time.

## Parameters 

### CardActions

| Property                     | Type           | Default                  | Description                                             |
|------------------------------|----------------|--------------------------|---------------------------------------------------------|
| width                        | double         | required                 | widget width                                            |
| height                       | double         | required                 | widget height                                           |
| cardCursor                   | MouseCursor    | SystemMouseCursors.basic | the type of cursor displayed when hovering on the card  |
| buttonsCursor                | MouseCursor    | SystemMouseCursors.basic | the type of cursor displayed when hovering over buttons |
| borderRadius                 | double         | 0                        | rounded card edges                                      |
| showToolTip                  | bool           | true                     | Show/Hide toolTips                                      |
| axisDirection                | CardActionAxis | CardActionAxis.bottom    | CardActions Axis Direction                              |
| splashColor                  | Color          | Colors.white24           | Splash color when hovering                              |
| backgroundColor              | Color          | Color(0xff242120)        | Background color                                        |
| closeCardWhenExecuteFunction | bool           | false                    | close the card when you press a button                  |

### CardActionButton

| Property           | Type         | Default  | Description                                       |
|--------------------|--------------|----------|---------------------------------------------------|
| icon               | Icon         | required | Button icon                                       |
| label              | String       | required | text to be displayed in the Tool Tip              |
| onPress            | VoidCallback | required | callback to be executed when you press the button |
| closeCardWhenPress | bool         | false    | when you press the button, the card closes.       |
| cursor             | MouseCursor? | null     | Cursor displayed when hovering on the button      |

CHANGELOG.md
## 0.0.1

* Initial public version.


test/card_actions_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:card_actions/card_actions.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/card_actions.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'card_action_button.dart';
import 'card_provider.dart';

enum CardActionAxis { right, bottom }

class CardActions extends StatelessWidget {
  const CardActions({
    super.key,
    required this.child,
    required this.width,
    required this.height,
    required this.actions,
    this.cardCursor = SystemMouseCursors.basic,
    this.buttonsCursor = SystemMouseCursors.basic,
    this.borderRadius = 0,
    this.showToolTip = true,
    this.axisDirection = CardActionAxis.bottom,
    this.splashColor = Colors.white24,
    this.closeCardWhenExecuteFunction = false,
    this.backgroundColor = const Color(0xff242120),
  });

  /// the type of cursor displayed when hovering on the card
  final MouseCursor cardCursor;

  /// the type of cursor displayed when hovering over buttons
  final MouseCursor buttonsCursor;

  /// Background color
  final Color backgroundColor;

  /// list of all the buttons that the card is going to have
  final List<CardActionButton> actions;

  /// your card
  final Widget child;

  /// widget width
  final double width;

  /// widget height
  final double height;

  /// the radius of the edge that the card is to have
  final double borderRadius;

  /// Splash color when hovering
  final Color splashColor;

  /// Show/Hide toolTips
  final bool showToolTip;

  /// close the card when you press a button
  final bool closeCardWhenExecuteFunction;

  /// CardActions Axis Direction
  final CardActionAxis axisDirection;

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider(
      create: (_) => CardProvider(),
      child: Builder(
        builder: (BuildContext context) {
          final cardProvider = Provider.of<CardProvider>(context);

          final transform = cardProvider.isHovered
              ? cardProvider.getMatrixTransform()
              : Matrix4.identity();

          cardProvider.splashColor = splashColor;
          cardProvider.buttonsCursor = buttonsCursor;
          cardProvider.showToolTip = showToolTip;
          cardProvider.whenCallFunctionCloseCard = closeCardWhenExecuteFunction;
          cardProvider.hoverDirection = axisDirection;
          cardProvider.heightCard = height;
          cardProvider.widthCard = width;
          return MouseRegion(
            cursor: cardCursor,
            onEnter: (_) {
              cardProvider.isHovered = true;
            },
            onExit: (_) {
              cardProvider.isHovered = false;
            },
            child: AnimatedContainer(
              width: cardProvider.width,
              height: cardProvider.height,
              duration: const Duration(milliseconds: 200),
              child: Stack(
                alignment: cardProvider.getAligment(),
                children: [
                  AnimatedContainer(
                    duration: const Duration(milliseconds: 200),
                    transform: transform,
                    child: Container(
                      width: width,
                      height: height,
                      decoration: BoxDecoration(
                        color: backgroundColor,
                        borderRadius: BorderRadius.circular(borderRadius),
                      ),
                      child: SizedBox(
                        height: height,
                        width: width,
                        child: cardProvider.isBottom()
                            ? _BottomWidget(
                                borderRadius: borderRadius,
                                width: width,
                                actions: actions,
                              )
                            : _RightWidget(
                                borderRadius: borderRadius,
                                height: height,
                                actions: actions,
                              ),
                      ),
                    ),
                  ),
                  child,
                ],
              ),
            ),
          );
        },
      ),
    );
  }
}

class _RightWidget extends StatelessWidget {
  const _RightWidget({
    required this.borderRadius,
    required this.height,
    required this.actions,
  });
  final double borderRadius;
  final double height;
  final List<CardActionButton> actions;

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.end,
      children: [
        ClipRRect(
          borderRadius: BorderRadius.only(
            topRight: Radius.circular(borderRadius),
            bottomRight: Radius.circular(borderRadius),
          ),
          child: Column(
            mainAxisAlignment: actions.length == 1
                ? MainAxisAlignment.center
                : MainAxisAlignment.spaceAround,
            children: [
              ...actions,
            ],
          ),
        ),
      ],
    );
  }
}

class _BottomWidget extends StatelessWidget {
  const _BottomWidget({
    required this.borderRadius,
    required this.width,
    required this.actions,
  });

  final double borderRadius;
  final double width;
  final List<CardActionButton> actions;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.end,
      children: [
        ClipRRect(
          borderRadius: BorderRadius.only(
            bottomLeft: Radius.circular(borderRadius),
            bottomRight: Radius.circular(borderRadius),
          ),
          child: Row(
            mainAxisAlignment: actions.length == 1
                ? MainAxisAlignment.center
                : MainAxisAlignment.spaceAround,
            children: [
              ...actions,
            ],
          ),
        ),
      ],
    );
  }
}


lib/card_provider.dart
import 'package:flutter/material.dart';
import 'card_actions.dart';

class CardProvider with ChangeNotifier {
  late Color splashColor;
  late bool showToolTip;
  late bool whenCallFunctionCloseCard;
  late CardActionAxis hoverDirection;
  late MouseCursor buttonsCursor;

  late double heightCard;
  late double widthCard;

  bool _isHovered = false;
  late double _height = heightCard;
  late double width = widthCard;
  double get height => _height;

  getMatrixTransform() {
    if (hoverDirection == CardActionAxis.bottom) {
      return Matrix4.identity()..setTranslationRaw(0, 50, 0);
    }

    if (hoverDirection == CardActionAxis.right) {
      return Matrix4.identity()..setTranslationRaw(50, 0, 0);
    }
  }

  bool isBottom() {
    if (hoverDirection == CardActionAxis.bottom) return true;

    return false;
  }

  getAligment() {
    if (hoverDirection == CardActionAxis.bottom) {
      return Alignment.topCenter;
    }

    if (hoverDirection == CardActionAxis.right) {
      return Alignment.centerLeft;
    }
  }

  bool get isHovered => _isHovered;
  set isHovered(bool value) {
    if (hoverDirection == CardActionAxis.bottom) {
      _isHovered = value;
      if (value) _height = _height + 50;
      if (!value) _height = _height - 50;
    }

    if (hoverDirection == CardActionAxis.right) {
      _isHovered = value;
      if (value) width = width + 50;
      if (!value) width = width - 50;
    }

    notifyListeners();
  }
}


lib/card_action_button.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'card_provider.dart';

class CardActionButton extends StatefulWidget {
  const CardActionButton({
    super.key,
    required this.icon,
    required this.label,
    required this.onPress,
    this.closeCardWhenPress = false,
    this.cursor,
  });

  /// Cursor displayed when hovering on the button
  final MouseCursor? cursor;

  /// Button icon
  final Icon icon;

  /// text to be displayed in the Tool Tip
  final String label;

  /// callback to be executed when you press the button
  final VoidCallback onPress;

  /// when you press the button, the card closes
  final bool closeCardWhenPress;

  @override
  State<CardActionButton> createState() => _CardActionButtonState();
}

class _CardActionButtonState extends State<CardActionButton> {
  bool buttonIsHovered = false;
  @override
  Widget build(BuildContext context) {
    final cardProvider = Provider.of<CardProvider>(context, listen: false);

    MouseCursor getCurrentCursor() {
      if (widget.cursor != null) {
        return widget.cursor!;
      }

      return cardProvider.buttonsCursor;
    }

    return Expanded(
      child: GestureDetector(
        onTap: () {
          widget.onPress();

          if (cardProvider.whenCallFunctionCloseCard ||
              widget.closeCardWhenPress) {
            cardProvider.isHovered = false;
          }
        },
        child: MouseRegion(
          cursor: getCurrentCursor(),
          onEnter: (event) {
            setState(() {
              buttonIsHovered = true;
            });
          },
          onExit: (event) {
            setState(() {
              buttonIsHovered = false;
            });
          },
          child: cardProvider.showToolTip
              ? Tooltip(
                  message: widget.label,
                  textStyle: const TextStyle(fontSize: 18, color: Colors.white),
                  padding: const EdgeInsets.symmetric(horizontal: 20),
                  decoration: BoxDecoration(
                    color: const Color(0xff242120),
                    borderRadius: BorderRadius.circular(15),
                  ),
                  height: 35,
                  verticalOffset: cardProvider.isBottom() ? -70 : 30,
                  child: _BodyButton(
                    buttonIsHovered: buttonIsHovered,
                    cardProvider: cardProvider,
                    widget: widget,
                  ),
                )
              : _BodyButton(
                  buttonIsHovered: buttonIsHovered,
                  cardProvider: cardProvider,
                  widget: widget,
                ),
        ),
      ),
    );
  }
}

class _BodyButton extends StatelessWidget {
  const _BodyButton({
    required this.buttonIsHovered,
    required this.cardProvider,
    required this.widget,
  });

  final bool buttonIsHovered;
  final CardProvider cardProvider;
  final CardActionButton widget;

  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      duration: const Duration(milliseconds: 200),
      width: cardProvider.isBottom() ? null : 50,
      height: 60,
      decoration: BoxDecoration(
        color: buttonIsHovered ? cardProvider.splashColor : Colors.transparent,
      ),
      child: widget.icon,
    );
  }
}


example/lib/main.dart
import 'package:card_actions/card_action_button.dart';
import 'package:card_actions/card_actions.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatefulWidget {
  const MainApp({super.key});

  @override
  State<MainApp> createState() => _MainAppState();
}

class _MainAppState extends State<MainApp> {
  int counter = 0;

  void increment() {
    setState(() {
      counter++;
    });
  }

  void decrement() {
    setState(() {
      counter--;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        backgroundColor: const Color(0xfff7f8fd),
        body: Center(
          child: CardActions(
            buttonsCursor: SystemMouseCursors.click,
            backgroundColor: const Color(0xffff7270),
            axisDirection: CardActionAxis.bottom,
            borderRadius: 15,
            width: 600,
            height: 420,
            actions: [
              CardActionButton(
                icon: const Icon(
                  Icons.add,
                  color: Colors.white,
                  size: 30,
                ),
                label: 'Increment',
                onPress: increment,
              ),
              CardActionButton(
                icon: const Icon(
                  Icons.restore_sharp,
                  color: Colors.white,
                  size: 30,
                ),
                label: 'Reset',
                onPress: () {
                  setState(() {
                    counter = 0;
                  });
                },
              ),
              CardActionButton(
                icon: const Icon(
                  Icons.remove,
                  color: Colors.white,
                  size: 30,
                ),
                label: 'Decrement',
                onPress: decrement,
              ),
            ],
            child: UserCard(counter: counter),
          ),
        ),
      ),
    );
  }
}

class UserCard extends StatelessWidget {
  const UserCard({
    super.key,
    required this.counter,
  });
  final int counter;

  @override
  Widget build(BuildContext context) {
    return Container(
      width: 600,
      height: 420,
      decoration: BoxDecoration(
        color: const Color.fromARGB(255, 51, 49, 49),
        borderRadius: BorderRadius.circular(15),
        boxShadow: const [
          BoxShadow(
            color: Colors.black12,
            offset: Offset(10, 10),
            blurRadius: 20,
            // spreadRadius: 20,
          )
        ],
      ),
      child: Padding(
        padding: const EdgeInsets.all(60),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            const Text(
              'Counter Example',
              style: TextStyle(
                color: Colors.white,
                fontWeight: FontWeight.bold,
                fontSize: 50,
              ),
            ),
            const SizedBox(height: 20),
            Text(
              counter.toString(),
              style: const TextStyle(
                color: Colors.white,
                fontWeight: FontWeight.bold,
                fontSize: 60,
              ),
            )
          ],
        ),
      ),
    );
  }
}
